tools/codegen/model_generator/model.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model_generator
+
+import (
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	tmplExtns "istio.io/mixer/tools/codegen/template_extension"
+	"github.com/golang/protobuf/proto"
+)
+
+/*
+Data model to be used to generate code for istio/mixer
+*/
+type Model struct {
+	// top level fields
+	Name        string
+	Check       bool
+	PackageName string
+	VarietyName string
+
+	// imports
+	Imports []string
+
+	// Constructor fields
+	ConstructorFields []FieldInfo
+}
+
+type FieldInfo struct {
+	Name string
+	Type TypeInfo
+}
+
+type TypeInfo struct {
+	Name string
+}
+
+const FullNameOfExprMessage = "istio_mixer_v1_config_template.Expr"
+const FullNameOfExprMessageWithPtr = "*" + FullNameOfExprMessage
+
+// Creates a Model object.
+func CreateModel(parser *FileDescriptorSetParser) (Model, error) {
+	model := &Model{}
+
+	templateProto, err := getTmplFileDesc(parser.allFiles)
+	if err != nil {
+		return *model, err
+	}
+	if !templateProto.proto3 {
+		return *model, fmt.Errorf("Only proto3 template files are allowed")
+	}
+	// set the current generated code package to the package of the
+	// templateProto. This will make sure references within the
+	// generated file into the template's pb.go file are fully qualified.
+	parser.packageName = goPackageName(templateProto.GetPackage())
+
+	err = model.addTopLevelFields(templateProto)
+	if err != nil {
+		return *model, err
+	}
+
+	// ensure Constructor is present
+	cnstrDesc, err := getRequiredMsg(templateProto, "Constructor")
+	if err != nil {
+		return *model, err
+	}
+
+	// ensure Constructor is present
+	_, err = getRequiredMsg(templateProto, "Type")
+	if err != nil {
```

Comment:
Here and elsewhere, combine these kinds of 'if' statements with the previous line.